### PR TITLE
Hackathon 5th No.63 PhyCRNet: Physics-informed Convolutional-Recurrent Network for Solving Spatiotemporal PDEs

### PR DESCRIPTION
**Pull request by co63oc**
### PR types

Others
### PR changes

Others
### Describe

https://github.com/PaddlePaddle/Paddle/issues/57262
image

Loss精度
| - | torch | paddle |
| - | - | - |
| loss | 18.55 | 17.86 |

运行 200 次截图
torch
image
paddle
image

### CODE REVIEW

**Comment by HydrogenSulfate** [comment]
黑客松的赛题有要求需要基于PaddleScience套件进行开发，
image

**Comment by co63oc** [comment]
> 黑客松的赛题有要求需要基于PaddleScience套件进行开发

图片
这个模型一个epoch多次调用model.forward，传递不同initial_state，现在Solver接口是每个epoch调用一次forward，不知道这是要用什么api转换

**Comment by HydrogenSulfate** [comment]
> > 黑客松的赛题有要求需要基于PaddleScience套件进行开发
> 
> 图片 这个模型一个epoch多次调用model.forward，传递不同initial_state，下一次的initial_state是上一次计算返回，现在Solver接口是每个epoch调用一次forward，不知道这是要用什么api转换

可以把内层的for循环逻辑写到output_transform里

**Comment by co63oc** [comment]
更新使用 PaddleScience套件，增加ppsci.arch.PhyCRNet

**Comment by co63oc** [comment]
没有使用ppsci.utils.load_mat_file，ppsci.utils.load_mat_file中会对数据reshape，没有明白怎么要对数据reshape
image

File: docs/zh/examples/phycrnet.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+``` py linenums="1" title="phycrnet"
+--8<--
+examples/phycrnet/main.py
+--8<--
+```
+
+## 5. 结果展示
+
+PhyCRNet 案例针对 epoch=200 和 learning\_rate=1e-4 的参数配置进行了实验，结果返回Loss为 17.86。
```
Comment (wangguan1995) [inline docs/zh/examples/phycrnet.md:119]: # compute the error
    error = frobenius_norm(np.array(ten_pred)-np.array(ten_true)) / frobenius_norm(
        np.array(ten_true))

    print('The predicted error is: ', error)

    u_pred = output[:-1, 0, :, :].detach().cpu().numpy()
    u_pred = np.swapaxes(u_pred, 1, 2) # [h,w] = [y,x]
    u_true = truth[:, 0, :, :]

    t_true = np.linspace(0, 2, 1001)
    t_pred = np.linspace(0, 2, time_steps)

    plt.plot(t_pred, u_pred[:, 32, 32], label='x=32, y=32, CRL')
    plt.plot(t_true, u_true[:, 32, 32], '--', label='x=32, y=32, Ref.')
    plt.xlabel('t')
    plt.ylabel('u')
    plt.xlim(0, 2)
    plt.legend()
    plt.savefig(fig_save_path + "x=32,y=32.png")
    plt.close("all")
    # plt.show()

    # plot train loss
    plt.figure()
    plt.plot(train_loss, label = 'train loss')
    plt.yscale('log')
    plt.legend()
    plt.savefig(fig_save_path + 'train loss.png', dpi = 300)

科学计算类论文复现，更关注 物理error和最后的物理场效果，参考原文代码复现error和对比图
（可以从checkpoint加载模型直接eval，不需要重新训练）

**Review by wangguan1995 — COMMENTED** [review]
need to align with paper

File: docs/zh/examples/phycrnet.md (hunk 1 above)
Comment (co63oc) [reply docs/zh/examples/phycrnet.md:119]: 增加变量graph，设置为True生成error，对比图
image
image

File: examples/phycrnet/main.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+        ten_true.append([u_star, v_star])
+        ten_pred.append([u_pred, v_pred])
+
+    # compute the error
+    error = functions.frobenius_norm(
+        np.array(ten_pred) - np.array(ten_true)
+    ) / functions.frobenius_norm(np.array(ten_true))
+
+    print("The predicted error is: ", error)
```
Comment (wangguan1995) [inline examples/phycrnet/main.py:171]: error要求达到：源码，论文（1e-3）取较大值

File: docs/zh/examples/phycrnet.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+训练过程会调用优化器来更新模型参数，此处选择 `Adam` 优化器并设定 `learning_rate` 为 1e-4。
+
+``` py linenums="242"
+--8<--
+examples/phycrnet/main.py:242:242
+--8<--
+```
+
+### 3.7 模型训练与评估
```
Comment (wangguan1995) [inline docs/zh/examples/phycrnet.md:91]: 按照
image
翻译，并介绍误差

File: examples/phycrnet/README.md
Comment (wangguan1995) [inline examples/phycrnet/README.md:1]: 删掉不需要

Comment (wangguan1995) [reply examples/phycrnet/README.md:1]: 有价值的部分合入docs

File: ppsci/arch/phycrnet.py [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+            [0, 0, -8 / 12, 0, 0],
+            [0, 0, 0, 0, 0],
+            [0, 0, 8 / 12, 0, 0],
+            [0, 0, -1 / 12, 0, 0],
+        ]
+    ]
+]
+
+# specific parameters for burgers equation
+def initialize_weights(module):
```
Comment (wangguan1995) [inline ppsci/arch/phycrnet.py:60]: 注释

File: ppsci/arch/phycrnet.py [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+                input_stride=[2, 2, 2, 1],
+                input_padding=[1, 1, 1, 1],
+                dt=0.002,
+                num_layers=[3, 1],
+                upscale_factor=8
+            )
+    """
+
+    def __init__(
+        self,
```
Comment (wangguan1995) [inline ppsci/arch/phycrnet.py:102]: type hint

File: ppsci/arch/phycrnet.py [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+        hidden_channels,
+        input_kernel_size,
+        input_stride,
+        input_padding,
+    ):
+        super(ConvLSTMCell, self).__init__()
+
+        self.input_channels = input_channels
+        self.hidden_channels = hidden_channels
+        self.hidden_kernel_size = 3
```
Comment (wangguan1995) [inline ppsci/arch/phycrnet.py:243]: 3是怎么来的?

File: ppsci/arch/phycrnet.py [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+        self.input_kernel_size = input_kernel_size
+        self.input_stride = input_stride
+        self.input_padding = input_padding
+        self.num_features = 4
```
Comment (wangguan1995) [inline ppsci/arch/phycrnet.py:247]: 4是怎么来的?

File: ppsci/arch/phycrnet.py [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+
+    def forward(self, input):
+        derivative = self.filter(input)
+        return derivative / self.resol
+
+
+class loss_generator(nn.Layer):
+    """Loss generator for physics loss"""
+
+    def __init__(self, dt=(10.0 / 200), dx=(20.0 / 128)):
```
Comment (wangguan1995) [inline ppsci/arch/phycrnet.py:469]: 默认值删掉

File: examples/phycrnet/burgers_data.py [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+    fig.colorbar(cf, ax=ax[1], fraction=0.046, pad=0.04)
+
+    plt.savefig(fig_save_dir + "/uv_[i=%d].png" % (num))
+    plt.close("all")
+
+
+if __name__ == "__main__":
+    # grid size
+    M, N = 128, 128
+    n_simu_steps = 30000
```
Comment (wangguan1995) [inline examples/phycrnet/burgers_data.py:234]: 40000

File: examples/phycrnet/main.py [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+    else:
+        import os
+
+        fig_save_path = "output/figures/"
+        if not os.path.exists(fig_save_path):
+            os.makedirs(fig_save_path, True)
+        layer_state_dict = paddle.load("output/phycrnet.pdparams")
+        model.set_state_dict(layer_state_dict)
+        model.register_output_transform(None)
+        output_graph(model, input_dict_val, fig_save_path)
```
Comment (wangguan1995) [inline examples/phycrnet/main.py:341]: input_dict_val进行修改，尝试完成时间变量t，从0s到4s的u_predict和u_true的对比，以及a-RMSE计算

**Review by wangguan1995 — COMMENTED** [review]
need a-RMSE plot for t in [0,4]

**Comment by co63oc** [comment]
暂时关闭
